031-statefulset-removal.md

```diff
@@ -0,0 +1,242 @@
+StrimziPodSet will be used only internally.
+Users will not be expected to interact with.
+
+The StrimziPodSets will have a following structure:
+* The usual metadata, kind and API identifications
+* A `.spec` section with:
+    * Pod selector to be able to find the pods managed by this StrimziPodSet
+    * List of Pods (full Pod YAML definition including metadata section, but without status section)
+* A status section with:
+    * Number of managed, current and ready Pods
+    * List of error conditions in case of problems
+    * Observed generation
+
+The StrimziPodSet will be used very similarly to the StatefulSets:
+* The Pods will be always created in the same namespace as the pod set
+* Garbage collection will be used to delete the Pods
+
+Instead of using a Pod templates, and generating Pods with fixed names, StrimziPodSet will use full Pod structures including custom names.
+The configuration retains the usefulness of a StatefulSet, but removes the limitations we have identified.
+
+The StrimziPodSet will be created by the `KafkaAssemblyOperator` class in the same place in the reconciliation as the StatefulSet.
+This is also where the StrimziPodSet will be patched when configuration changes.
+
+A new controller will be developed to reconcile the StrimziPodSet resources.
+It will run as part of the Strimzi cluster operator container.
+It will follow a more lightweight controller approach rather than the long-running _heavy_ approach of our current assembly operators.
+It will be designed to iterate quickly rather than wait.
+It will not wait for Pod readiness.
+It will consume events about the StrimziPodSets and their Pods and react to them.
+It will run in its own thread and use a work queue to queue the events and the reconciliations.
+
+![Controller design](images/031-StrimziPodSet-controller.png)
+
+It will be responsible for:
+* Creating pods which do not exist
+* Deleting old pods removed from the StrimziPodSet
+* Update the StrimziPodSet status subresource
+
+When a new StrimziPodSet is created, it will create its pods.
+When a StrimziPodSet is modified, it will check if any pods need to be added or deleted.
+When a Pod is deleted or modified, it will identify the StrimziPodSet it belongs to and reconcile the whole StrimziPodSet.
+
+![StrimziPodSet hierarchy](images/031-strimzi-with-StrimziPodSet.png)
+
+The StrimziPodSet controller will not be responsible for handling the rolling updates.
+They will continue to be handled in the same way by the `KafkaAssemblyOperator` and in the accompanied classes such as `KafkaRoller` and `ZooKeeperRoller`.
+They will identify which pods need to be rolled and delete them.
+The only difference will be that instead of the Kubernetes StatefulSet controller, it will be the StrimziPodSet controller which will create a new pod.
+For each Pod managed by a StrimziPodSet, Strimzi will calculate a revision (really a _hash_ of the Pod configuration).
```

Comment:
Rather than "Strimzi will calculate" it would be better to say which component would do this (since we're talking about several in this proposal)

Reply:
Improved.